### DISCUSSION TITLE
Expose consumerTag during Consumer_Shutdown event

### DISCUSSION THREAD

**Discussion by plewam** [rabbitmq/rabbitmq-dotnet-client#1433] [Ideas]
Hi All,

I subscribe to the `Shutdown` event of an `EventingBasicConsumer`. However when I receive the event I cannot track back which consumer was cancelled as I don't have the consumer tag available within the `ShutdownEventArgs`. This way I cannot evaluate the shutdown reason and eventually reconnect. Would it be possible to add the consumerTag?

I use the a manual ack of rmq messages as a method of throttling if my consumer is not able to process the messages quick enough into a sink (Database in this scenario).
After 30 minutes, rmq cancels my consumer if I have not ack'ed the message. In case of a database maintenance, restart this might happen. In this scenario I would like to reconnect... But I don't know which consumer to reconnect...

On the other hand I also receive the `ConsumerCancelled` event, but this one **only** contains the `consumerTag` and not the shutdown reason.

**Comment by michaelklishin** [comment]
@plewam are you sure you want to "connect" manually (that is, roll your own [connection recovery](https://www.rabbitmq.com/dotnet-api-guide.html#recovery))?

If you do and realize the amount of effort it will take, please keep in mind that connection recovery is not initiated by consumers. It is initiated by something that reacts to **connection** shutdown events.

**Comment by lukebakken** [comment]
@plewam this seems like a reasonable request. I will see if the Java client provides this information (just out of curiosity).

@michaelklishin's point is valid - connection and topolgy recovery is already implemented in this client library. Have you tried it and found it insufficient for your use case?

**Reply by michaelklishin** [reply]
FTR, I am not against adding a new field to the consumer-specific version of `ShutdownEventArgs` but initiating recovery from a consumer event handler sounds backwards to what our client libraries have been doing (or recommending to do as examples) since 2013.

**Reply by lukebakken** [reply]
I'm hoping the plan is to _not_ do recovery from an event handler, because that would be fraught with peril - cc @plewam

**Reply by plewam** [reply]
I may have explained my problem in a wrong way. If you consume from a queue with manual acknowledgment the consumer is getting cancelled by the peer after 30 minutes. This only happens if no acks are done within that time for that consumer. This however might be the case in my scenario. For all the other uses cases the client side auto recovery is working and I am in fact using it.

**Reply by michaelklishin** [reply]
For tracking consumer cancellations and re-registering a consumer this would be a perfectly valid addition.

**Reply by plewam** [reply]
I mean if the library could also recover my described consumer disconnect, that would be great as well. Then I don't have to do the work :)

**Reply by michaelklishin** [reply]
The library will not decide what to do with your cancelled consumers. It does not have enough context. We explicitly state in the connection recovery docs that it is an opinionated process that works for most people but may or may not work for you.

**Reply by michaelklishin** [reply]
For example, if a queue is permanently deleted, trying to re-register (with or without a re-declaration of the queue) may be the exact opposite of what you want.

We have plenty of examples of applications that were designed to retry everything and one day, you just want them to do nothing for a period of time, and the only way is to stop them, which may or may not be easy to achieve.

"Be careful about what you wish for" and all.

**Reply by plewam** [reply]
Alright, your argument is valid. I can do the recover on my own then if I got the consumerTag available.

**Reply by lukebakken** [reply]
Yes, this library recovers consumers but my guess is that it may not do it in the case of a consumer timeout.

The root cause of this issue is that you are holding on to a message for too long, or taking too long to process it. Can you fix the root cause by modifying your workflow for messages that may take too long to process?

The next fix would be to increase the consumer timeout value. RabbitMQ 3.12 and later allows this to be set per-queue: https://rabbitmq.com/consumers.html#acknowledgement-timeout

You can also set it per-consumer via the `x-consumer-timeout` consumer argument ([code](https://github.com/rabbitmq/rabbitmq-server/pull/7981/files#diff-7edeaf2557dcfdaae22493ecae0f85de7ca9b85fdb38c8cc63c7c41b7e1acde4R35)). This somehow did not get documented and I'll get right on that.

**Reply by plewam** [reply]
Yes it may be because of my implementation. I use rmq as the source of data and a database connector as the sink. If my sink (database) is unable to process the data (e.g. due to the database being down, or updates) I risk flooding my application with data that cannot go anywhere. Therefore I throttle down the ack's of messages. However in my above described scenario I stop ack's completely. The only thing I then can do is to restart the application once every consumer was cancelled. I know kind of try to programm around that. However I got not other idea.

**Reply by lukebakken** [reply]
Never mind, `x-consumer-timeout` was removed in https://github.com/rabbitmq/rabbitmq-server/pull/8260

@michaelklishin @deadtrickster - do you remember why the above was removed? I'm just curious.

**Reply by michaelklishin** [reply]
We didn't see the value in having not just per queue but per consumer values. How common is it for N consumers on the same queue to have different processing timeouts?

**Comment by lukebakken** [comment]
Hello @plewam -

I will have a new alpha release for you to test soon.

With regard to `EventingBasicConsumer`, when the `Shutdown` event is raised, the `sender` argument will be the instance of the `EventingBasicConsumer` that is raising the event, and this instance will have a `ConsumerTags[]` property where you can find the consumer tags for the event.

I'm going to close this discussion, but let me know if this meets your requirements.

**Reply by plewam** [reply]
@lukebakken Perfect, thats exactly what I need. Thank you
